src/sema/variables.rs

```diff
@@ -460,6 +461,143 @@ pub fn variable_decl<'a>(
+                        .collect();
+
+                    body = vec![
+                        Statement::VariableDecl(pt::Loc::Implicit, var_no, param, Some(expr)),
+                        Statement::Return(
+                            pt::Loc::Implicit,
+                            Some(Expression::List {
+                                loc: pt::Loc::Implicit,
+                                list,
+                            }),
+                        ),
+                    ];
+
+                    returns = def.fields.clone();
+                } else {
+                    let ty = Type::StorageRef(false, ty.clone().into());
+
+                    let var_no = symtable
+                        .add(
+                            &pt::Identifier {
+                                loc: pt::Loc::Implicit,
+                                name: "_".into(),
+                            },
+                            ty.clone(),
+                            ns,
+                            VariableInitializer::Solidity(Some(expr.clone())),
+                            VariableUsage::LocalVariable,
+                            Some(pt::StorageLocation::Storage(pt::Loc::Implicit)),
+                        )
+                        .unwrap();
+
+                    symtable.vars[&var_no].read = true;
+                    symtable.vars[&var_no].assigned = true;
+
+                    let def = str.definition(ns);
+
+                    let list = def
+                        .fields
+                        .iter()
+                        .enumerate()
+                        .map(|(i, param)| Expression::StorageLoad {
+                            loc: pt::Loc::Implicit,
+                            ty: param.ty.clone(),
+                            expr: Expression::StructMember {
+                                loc: pt::Loc::Implicit,
+                                ty: Type::StorageRef(false, def.fields[i].ty.clone().into()),
+                                expr: Expression::Variable {
+                                    loc: pt::Loc::Implicit,
+                                    ty: ty.clone(),
+                                    var_no,
+                                }
+                                .into(),
+                                field: i,
+                            }
+                            .into(),
+                        })
+                        .collect();
+
+                    body = vec![
+                        Statement::VariableDecl(pt::Loc::Implicit, var_no, param, Some(expr)),
+                        Statement::Return(
+                            pt::Loc::Implicit,
+                            Some(Expression::List {
+                                loc: pt::Loc::Implicit,
+                                list,
+                            }),
+                        ),
+                    ];
+
+                    returns = def.fields.clone();
+                }
```

Comment:
The two branches of `if constant { .. } else {...}` have nearly identical code. You can simplify most of it.

Reply:
Your new version looks much cleaner. Thanks!